lib/ntlmauth/ntlmauth.cc

```diff
@@ -108,10 +108,19 @@ ntlm_fetch_string(const ntlmhdr *packet, const int32_t packet_size, const strhdr
     int32_t o = le32toh(str->offset);
     // debug("ntlm_fetch_string(plength=%d,l=%d,o=%d)\n",packet_size,l,o);
 
-    if (l < 0 || l > NTLM_MAX_FIELD_LENGTH || o + l > packet_size || o == 0) {
-        debug("ntlm_fetch_string: insane data (pkt-sz: %d, fetch len: %d, offset: %d)\n", packet_size,l,o);
+    if (l < 0 || NTLM_MAX_FIELD_LENGTH < l) {
+        debug("ntlm_fetch_string: insane string length (pkt-sz: %d, fetch len: %d, offset: %d)\n", packet_size,l,o);
         return rv;
     }
+    else if (o <= 0 || packet_size < o) {
```

Comment:
IMHO, the following order is more natural, easier to read/grok:

```suggestion
    else if (o <= 0 || o > packet_size) {
```

The above "item > limit" order is also more consistent with the last `l` check below.

I also find it odd that offset `o` cannot be zero, but length `l` can. If anything, I would expect the opposite, but I know the official code rejects zero offsets, and I do not even pretend to understand this code.

Reply:
There is a prefix octet at offset 0. It cannot be part of any contained string. Strings can be 'empty' though since some are mandatory fields.